music21/analysis/enharmonics.py

```diff
@@ -0,0 +1,187 @@
+# Purpose:      Tools for returning best enharmonics
+#
+# Authors:      Mark Gotham
+#               Michael Scott Cuthbert
+#
+# Copyright:    Copyright © 2017 Michael Scott Cuthbert and the music21 Project
+# License:      LGPL or BSD, see license.txt
+#-------------------------------------------------------------------------------
+
+import unittest
+import itertools
+from music21 import common
+from music21 import exceptions21
+from music21 import pitch
+from music21 import chord
+from music21 import musedata
+
+from music21 import environment
+_MOD = 'analysis.enharmonics'
+environLocal = environment.Environment(_MOD)
+
+
+class EnharmonicsException(exceptions21.Music21Exception):
+    pass
+
+class EnharmonicScoreRules:
+    def __init__(self):
+        self.sameStaffLine = False
+        self.alterationPenalty = 4
+        self.augDimPenalty = 2
+        self.mixSharpsFlatsPenalty = False
+
+class ChordEnharmonicScoreRules(EnharmonicScoreRules):
+    def __init__(self):
+        super().__init__()
+        self.mixSharpsFlatsPenalty = 2
+
+class EnharmonicSimplifier:
+    '''
+    Takes any pitch list input and returns the best enharmonic respelling according to the input
+    criteria and rule weightings.
+    Those criteria and rule weightings are currently fixed, but in future the user should be able
+    to select their own combination and weighting of rules according to preferences,
+    with predefined defaults for melodic and harmonic norms.
+    Note: EnharmonicSimplifier itself returns nothing.
+    '''
+    def __init__(self, pitchList, ruleClass=EnharmonicScoreRules):
+        if isinstance(pitchList[0], str):
+            pitchList = [pitch.Pitch(p) for p in pitchList]
+
+        self.pitchList = pitchList
+        self.ruleObject = ruleClass()
+        self.getRepresentations()
+
+    def getRepresentations(self):
+        '''
+        Takes a list of pitches or pitch names and retrieves all enharmonic spellings.
+        Note: getRepresentations itself returns nothing.
+        '''
+        allSpellings = []
+        for p in self.pitchList:
+            spellings = [p] + p.getAllCommonEnharmonics(1)
+            allSpellings.append(spellings)
+        self.allSpellings = allSpellings
+
+    def getProduct(self):
+        self.allPossibleSpellings = list(itertools.product(*self.allSpellings))
+
+    def bestPitches(self):
+        '''
+        Returns a list of pitches in the best enharmonic spelling according to the input criteria.
+        >>> pList1 = [pitch.Pitch('C'), pitch.Pitch('D'), pitch.Pitch('E')]
+        >>> es = analysis.enharmonics.EnharmonicSimplifier(pList1)
+        >>> es.bestPitches()
+        (<music21.pitch.Pitch C>, <music21.pitch.Pitch D>, <music21.pitch.Pitch E>)
+        >>> pList2 = ['D--', 'E', 'F##']
+        >>> es = analysis.enharmonics.EnharmonicSimplifier(pList2)
+        >>> es.bestPitches()
+        (<music21.pitch.Pitch C>, <music21.pitch.Pitch E>, <music21.pitch.Pitch G>)
+        '''
```

Comment:
What happens if there's also a C# in here -- will it return C,C#,E,G or C,Db,E,G?

Reply:
C# in that case to minimise the aug/ dim score (it depends on the order). What we have works well. but we need to run an experiment on highly-trained musical subjects to try and pin down  good, exact, starting values for this metric. I'll see if I can recruit someone.